src/Stack/Dot.hs

```diff
@@ -275,24 +275,37 @@ createDepLoader sourceMap installed globalDumpMap globalIdMap loadPackageDeps pk
                     ghcOptions = cpGhcOptions common
                 assert (pkgName == name) (loadPackageDeps pkgName version loc flags ghcOptions)
               Nothing ->
-                pure (Set.empty, payloadFromInstalled (Map.lookup pkgName installed))
+                -- If package is in global dump map. load deps from there (#4324)
+                case Map.lookup pkgName globalDumpMap of
+                  Nothing ->
+                    pure (Set.empty, payloadFromInstalled (Map.lookup pkgName installed))
+                  dp ->
+                    getDepsFromDump dp
       -- For wired-in-packages, use information from ghc-pkg (see #3084)
-      else case Map.lookup pkgName globalDumpMap of
-          Nothing -> error ("Invariant violated: Expected to find wired-in-package " ++ packageNameString pkgName ++ " in global DB")
-          Just dp -> pure (Set.fromList deps, payloadFromDump dp)
-            where
-              deps = map (\depId -> maybe (error ("Invariant violated: Expected to find " ++ ghcPkgIdString depId ++ " in global DB"))
-                                          Stack.Prelude.pkgName
-                                          (Map.lookup depId globalIdMap))
-                         (dpDepends dp)
-  where
+      else
+          getDepsFromDump $ Map.lookup pkgName globalDumpMap
+    where
     payloadFromLocal pkg = DotPayload (Just $ packageVersion pkg) (Just $ packageLicense pkg)
     payloadFromInstalled maybePkg = DotPayload (fmap (installedVersion . snd) maybePkg) $
         case maybePkg of
             Just (_, Library _ _ mlicense) -> mlicense
             _ -> Nothing
     payloadFromDump dp = DotPayload (Just $ pkgVersion $ dpPackageIdent dp) (Right <$> dpLicense dp)
 
+    getDepsFromDump dump =
+      case dump of
+          Nothing ->
+            let errText = "Invariant violated: Expected to find wired-in-package "
+            in error (errText ++ packageNameString pkgName ++ " in global DB")
```

Comment:
This error branch is actual only for wired-in packages but it potentially could also get applied for non-wired-in ones. I think it's better to to move it out of this function and make this function accept `DumpPackage` without `Maybe`, probably renaming it to something like `dumpPackageDeps` (though it doesn't sound like `createDeplLoader` choice of which I'm not sure I quite understand).

Reply:
Actually I'm not sure that branching on wired-in packages makes much sense, what do you think about refactoring the code of `createDepLoader` into something like `projectPackgeDeps <|> dependencyDeps <|> globalDeps` where each of those will do a lookup into a corresponding table?

Reply:
I was thinking about that myself. So would we just totally get rid of the `createDepLoader` function? And the `globalDeps` function would handle the case of wired in packages and global packages?

Reply:
Oops, somehow I didn't get this comment into my inbox. Getting rid of `createDepLoader` looks like a sane thing to me. In principle the whole idea is just a do a dependency traversal and the idea of some special "loaders" looks like there is something else but I guess there isn't. But this kind of refactoring doesn't looks to be required for #4324 so probably could be done in a separate PR if you want to follow that route.